### PR TITLE
MyBillsViewController

### PR DESCRIPTION
**Pull request by wrumble**
This PR adds an iCarousel(basically a 3D collectionview) that loads a users bills from directly from firebase.

 I added a lot of new code so I currently haven't set it up to test what happens when there are multiple bills or the contents of each carousel cell as that will contain a tableview, which i will probably apply learnings from this PR.

All tests added are mostly ui based apart from a couple unit tests added to Billtests.

 The delete button doesnt have any tests yet either as it will be better tested when more that one carousel cell exists.

Working on the conflicts now.

PR adds:

- Add and delete buttons code as well as visibility tests.
- Title label code and visibility tests.
- iCarousel code and visibilty tests, for one cell only.
- Updated BillTests for retrieving a signed in users bills.
- Adds code and test for segue-ing to NewBillViewController from add button

No bills
One bill

### CODE REVIEW

**Comment by yvettecook** [comment]
Screenshots for context would be 💯 , and also bullet points for what this PR adds? (looking at the diff seems to be a quite a few changes in here)

File: Splitter/EarlGreyWelcomeScreenTests/EarlGreyWelcomeScreenTests.swift [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        
+        register(email: registeredEmail,
+                 password: password,
+                 confirmationPassword: password)
+        
+        let appearedSuccesfully = waitForSuccess(of: assertion,
+                                                 with: toastLabel,
+                                                 conditionName: conditionName)
+        
+        GREYAssertTrue(appearedSuccesfully, reason: "Toast appeared with correct text")
```
Comment (amlcurran) [inline Splitter/EarlGreyWelcomeScreenTests/EarlGreyWelcomeScreenTests.swift:138]: When we were using Earl Grey at another client, we wrote this wrapper to make the tests a little more succicnt. Perhaps you might want to take a look at it in other PRs - I'd be interested to see how you find it!

https://github.com/novoda/ios-demos/blob/master/Earl-Grey-Demo/Earl-Grey-DemoTests/UITest.swift

File: Splitter/Splitter/Controllers/MyBillsViewController.swift [hunk 2]
```diff
@@ -8,20 +8,129 @@
...
     
-    var titleLabel = UILabel()
-    var currentUser: SplitterUser!
+    private let addButton = AddButton(accessID: AccesID.addButton)
+    private let deleteButton = DeleteButton(accessID: AccesID.deleteButton)
+
+    private let noBillsLabel = NoBillsLabel()
+    private var titleLabel = TitleLabel()
+    private var carousel = iCarousel()
+    private var carouselDatasource = BillCarouselDatasource()
```
Comment (amlcurran) [inline Splitter/Splitter/Controllers/MyBillsViewController.swift]: these can be `let`, right?

File: Splitter/Splitter/Controllers/MyBillsViewController.swift [hunk 3]
```diff
@@ -8,20 +8,129 @@
...
+    
+    private func hideNoBillsLabel() {
+        DispatchQueue.main.async { [weak self] in
+            self?.noBillsLabel.isHidden = true
+        }
+    }
+    
+    @objc private func addButtonWasTapped() {
+        let newBillViewController = NewBillViewController()
+        newBillViewController.currentUserID = currentUser.id
```
Comment (amlcurran) [inline Splitter/Splitter/Controllers/MyBillsViewController.swift]: do you only create the new bill view controller in code? 

If you do, you could perhaps make the `currentUserID` a initialiser parameter. This has the benefit of a) then being a `let` variable, and b) makes it more obvious that this class can't function without a currentUser. 

Whilst setting variables on a viewcontroller is a _very_ common pattern on iOS, it is a bit of an antipattern. The reason being that I have to know/remember to set the variable, but as it is just a random variable, I'm given no information that it is required. If it is an initialiser param, then I am given no choice but to add it - highlighting to the consumer of the class that it is a requirement. The posh name for this is "[connascence by algorithm](https://en.wikipedia.org/wiki/Connascence_(computer_programming)#Connascence_of_Algorithm_.28CoA.29)" - you have to know how the class functions in order to be able to use it.

File: Splitter/Splitter/Controllers/WelcomeScreenViewController.swift [hunk 4]
```diff
@@ -13,9 +13,9 @@ class WelcomeScreenViewController: UIViewController {
     
     private var currentUser: SplitterUser? {
         didSet {
-            let nextViewController = MyBillsViewController()
-            nextViewController.currentUser = self.currentUser
-            present(nextViewController, animated: true)
+            let myBillsViewController = MyBillsViewController()
+            myBillsViewController.currentUser = self.currentUser
```
Comment (amlcurran) [inline Splitter/Splitter/Controllers/WelcomeScreenViewController.swift]: see above - this could be a constructor param

**Review by amlcurran — APPROVED** [review]
Couple of comments, but I like all the UI tests 💯

File: Splitter/Splitter/Controllers/MyBillsViewController.swift (hunk 3 above)
Comment (wrumble) [reply Splitter/Splitter/Controllers/MyBillsViewController.swift]: This is great! Thanks. I dont know why i never thought of using an initialiser for the viewcontroller. Seems ridiculous now. :)

File: Splitter/EarlGreyWelcomeScreenTests/EarlGreyWelcomeScreenTests.swift (hunk 1 above)
Comment (wrumble) [reply Splitter/EarlGreyWelcomeScreenTests/EarlGreyWelcomeScreenTests.swift:138]: Definitely 👍

**Review by amlcurran — APPROVED** [review]

File: Splitter/Splitter/Controllers/MyBillsViewController.swift (hunk 2 above)
Comment (wrumble) [reply Splitter/Splitter/Controllers/MyBillsViewController.swift]: Yep :) e9e96da7da9fa6c56336e41714f78e8f55f067fd

File: Splitter/Splitter/Controllers/MyBillsViewController.swift (hunk 3 above)
Comment (wrumble) [reply Splitter/Splitter/Controllers/MyBillsViewController.swift]: Done e9e96da7da9fa6c56336e41714f78e8f55f067fd

File: Splitter/Splitter/Controllers/WelcomeScreenViewController.swift (hunk 4 above)
Comment (wrumble) [reply Splitter/Splitter/Controllers/WelcomeScreenViewController.swift]: done e9e96da7da9fa6c56336e41714f78e8f55f067fd

**Comment by amlcurran** [comment]
Nice 💯

**Pull request merged**